src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+            {
+                _referenceType = referenceType;
+                _schema = schema;
+                _referenceName = referenceName;
+                _itemSpecification = itemSpecification;
+            }
+
+            public abstract Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference);
+
+            public IProjectRuleSnapshot GetProjectSnapshot(ConfiguredProject selectedConfiguredProject)
+            {
+                IProjectSubscriptionService? serviceSubscription = selectedConfiguredProject.Services.ProjectSubscription;
+                Assumes.Present(serviceSubscription);
+
+                serviceSubscription.ProjectRuleSource.SourceBlock.TryReceive(filter, out IProjectVersionedValue<IProjectSubscriptionUpdate> item);
+
+                _snapshotReferences = item.Value.CurrentState[_schema];
+
+                return _snapshotReferences;
+            }
+
+            private static bool filter(IProjectVersionedValue<IProjectSubscriptionUpdate> obj)
+            {
+                return true;
+            }
+
+            public List<ProjectSystemReferenceInfo> GetReferences()
+            {
+                List<ProjectSystemReferenceInfo> references = new List<ProjectSystemReferenceInfo>();
+
+                foreach (var item in _snapshotReferences.Items)
+                {
+                    string treatAsUsed = GetAttributeTreatAsUsed(item);
+                    string itemSpecification = GetAttributeItemSpecification(item);
+
+                    references.Add(new ProjectSystemReferenceInfo(_referenceType, itemSpecification, treatAsUsed == "True"));
+                }
+
+                return references;
+            }
+
+            private string GetAttributeTreatAsUsed(KeyValuePair<string, IImmutableDictionary<string, string>> item)
+            {
+                item.Value.TryGetValue("TreatAsUsed", out string treatAsUsed);
+                treatAsUsed = string.IsNullOrEmpty(treatAsUsed) ? "False" : treatAsUsed;
+
+                return treatAsUsed;
+            }
+
+            private string GetAttributeItemSpecification(KeyValuePair<string, IImmutableDictionary<string, string>> item)
+            {
+                item.Value.TryGetValue(_itemSpecification, out string itemSpecification);
+
+                return itemSpecification;
+            }
+
+            public async Task<bool> UpdateReferenceAsync(ConfiguredProject activeConfiguredProject, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+            {
+                bool wasUpdated = false;
+
+                string projectPath = activeConfiguredProject.UnconfiguredProject.FullPath;
+
+                string newValue = referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ? "True" : "False";
+
+                await activeConfiguredProject.Services.ProjectLockService.WriteLockAsync(async access =>
```

Comment:
@Lifengl What's the easiest way to set metadata on a IProjectItem? 

I don't think below is going to handle situations such as when the item comes from an import.

Reply:
I think perhaps we need a way to set arbitrary metadata here on an IUnresolvedReference.

Reply:
done. replaced with SetPropertyValueAsync